libtenzir/builtins/operators/parse.cpp

```diff
@@ -0,0 +1,213 @@
+        if (not parser_name_) {
+          diagnostic::error("expected parser name")
+            .primary(p.current_span())
+            .throw_();
+        }
+      } else {
+        parser_name_ = std::move(into_or_name);
+      }
+      TENZIR_ASSERT_CHEAP(parser_name_);
+      parser = plugins::find<parser_parser_plugin>(parser_name_->inner);
+      if (not parser) {
+        diagnostic::error("parser `{}` was not found", parser_name_->inner)
+          .primary(parser_name_->source)
+          .hint("must be one of: {}",
+                fmt::join(std::views::transform(
+                            collect(plugins::get<parser_parser_plugin>()),
+                            &parser_parser_plugin::name),
+                          ", "))
+          .throw_();
+      }
+    } catch (diagnostic& d) {
+      std::move(d)
+        .modify()
+        .usage("parse <input> [--into <output>] <parser> <args>...")
+        .docs("https://docs.tenzir.com/next/operators/transformations/parse")
+        .throw_();
+    }
+    TENZIR_ASSERT_CHEAP(parser);
+    parser_ = parser->parse_parser(p);
+    TENZIR_ASSERT_CHEAP(parser_);
+  }
+
+  auto
+  operator()(generator<table_slice> input, operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    for (auto&& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      auto batch = to_record_batch(slice);
+      auto schema = caf::get<record_type>(slice.schema());
+      auto index = schema.resolve_key(input_.inner);
+      if (not index) {
+        diagnostic::error("could not resolve `{}` for schema `{}`",
+                          input_.inner, slice.schema())
+          .primary(input_.source)
+          .emit(ctrl.diagnostics());
+        co_return;
+      }
+      auto [ty, array] = index->get(slice);
+      if (ty.kind().is_not<string_type>()) {
+        diagnostic::error("expected `string`, but got `{}` for schema `{}`",
+                          ty.kind(), slice.schema())
+          .primary(input_.source)
+          .emit(ctrl.diagnostics());
+        co_return;
+      }
+      auto string_array = std::dynamic_pointer_cast<arrow::StringArray>(array);
+      TENZIR_ASSERT_CHEAP(string_array);
+      auto results = parser_->parse_strings(string_array, ctrl);
+      auto total = int64_t{0};
+      for (auto& [result_ty, result] : results) {
+        TENZIR_ASSERT_CHEAP(result_ty.to_arrow_type()->Equals(result->type()));
+        total += result->length();
+      }
+      TENZIR_ASSERT_CHEAP(total == string_array->length());
+      const auto children
+        = batch->ToStructArray().ValueOrDie()->Flatten().ValueOrDie();
+      auto next = int64_t{0};
+      for (auto& [result_ty, result] : results) {
+        auto new_fields = collect(schema.fields());
+        // TODO: Some fields could be overwritten. We currently do not check for
+        // this, which can lead to having the multiple fields with the same name.
```

Comment:
One alternative to this is to (by default) replace the input field with the parsed output. Then you don't run into this problem at all, and don't need to handle this as a one-off for this operator.

Reply:
Yes, I have also thought about that. We can do it that way, but (1) it seems like inserting into the root record is a common choice, and (2) this would not work if the input is not a field (although it currently must be one). But yes, I'm still very much open to changing this. @eliaskosunen What do you think makes most sense?

Reply:
We decided to change this after some user feedback.